server/src/main/java/org/elasticsearch/search/profile/query/ProfileCollectorManager.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.profile.query;
+
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.CollectorManager;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link CollectorManager} that takes another CollectorManager as input and wraps all Collectors generated by it
+ * in an {@link InternalProfileCollector}. It delegates all the profiling to the generated collectors via {@link #getCollectorTree()}
+ * and joins them up when its {@link #reduce} method is called. The profile result can
+ */
+public final class ProfileCollectorManager implements CollectorManager<InternalProfileCollector, Void> {
+
+    private final CollectorManager<Collector, ?> collectorManager;
+    private CollectorResult collectorTree;
+
+    @SuppressWarnings("unchecked")
+    public ProfileCollectorManager(CollectorManager<? extends Collector, ?> collectorManager) {
+        this.collectorManager = (CollectorManager<Collector, ?>) collectorManager;
+    }
+
+    @Override
+    public InternalProfileCollector newCollector() throws IOException {
+        return new InternalProfileCollector(collectorManager.newCollector(), CollectorResult.REASON_SEARCH_TOP_HITS);
+    }
+
+    public Void reduce(Collection<InternalProfileCollector> profileCollectors) throws IOException {
+        List<Collector> unwrapped = profileCollectors.stream()
+            .map(InternalProfileCollector::getWrappedCollector)
+            .collect(Collectors.toList());
+        collectorManager.reduce(unwrapped);
+
+        List<CollectorResult> resultsPerProfiler = new ArrayList<>(profileCollectors.size());
+        long maxTime = 0;
+        for (InternalProfileCollector ipc : profileCollectors) {
+            CollectorResult ct = ipc.getCollectorTree();
+            resultsPerProfiler.add(ct);
+            if (maxTime < ct.getTime()) {
+                maxTime = ct.getTime();
+            }
+        }
+        this.collectorTree = new CollectorResult(
+            this.getClass().getSimpleName(),
+            CollectorResult.REASON_SEARCH_TOP_HITS + "_max",
```

Comment:
what is the rationale of tracking max? What are our options? I was wondering if we want to output the full tree structure of the collector manager, for each slice collector.

Reply:
Also the reason should be a constructor argument?

Reply:
> what is the rationale of tracking max? What are our options?
I thought "max" is most natural. The earlier child profilers (one per thread) should account for "search_top_hits" time spent in each thread. Since they run simultaneously, the maximum value should be an upper bound for the threads as a whole. Alternatives would be to use "sum" I guess, but that would hide the benefits of the parallel execution and appear longer than execution actually is?

> Also the reason should be a constructor argument?
I don't fully understand the question in this context

Reply:
Reason: it is hardcoded, which works well for dfs, but other phases will need to provide it, I thought.

As for what to output, I am thinking that we should not aggregate the output of the different collectors together, but rather expose them as they are, instead of having one collector with children, in a tree structure, we'll have a collector manager, which holds a list of collectors each with children, in a tree structure. How does that sound?